File: src/tests/1 Factorial/main.js

var N,          // input
    T,          // time limit (seconds)
    stopTime,   // time when we should stop
    iterations, // amount of finished cycles of computing

    res;        // current value

if (process.argv.length != 3) {
  process.stdout.write("Time limit parameter required:\n node main.js <time>\n");
} else {
  iterations = 0;
  T = Number.parseInt(process.argv[2], 10);
  process.stdout.write("Calculating factorial of N many times during "+T+" seconds\n");
  process.stdout.write("Input N:\n");
  
  process.stdin.setEncoding('utf8');
  process.stdin.on('readable', function() {
    var chunk = process.stdin.read();
    if (chunk != null) {
      if (typeof N === 'undefined') {
        N = Number.parseInt( chunk.replace("\n", ""), 10);
	setTimeout(start, 1);
      }
    }
  });
}

function start() {
  stopTime = (new Date()).getTime() + T * 1000; // NOW + T seconds
  while ( (new Date()).getTime() < stopTime) {
    reinit();
    calc(N);
    iterations++;
  }
  iterations--; // last iteration started before time was up but finished after time was up
  // required '>' mark for automatic benchmark
  process.stdout.write("> "+iterations+" iterations\n");
  process.stdout.write("Resulting "+N+"! = "+res+"\n");
  process.exit();
}

function reinit () {
  res = 1;
}

function calc(N) {
  for (var i = N; i > 1; i--) {
    res *= i;
  }
}

File: src/tests/1 Factorial/main.c
#include <stdio.h>
#include <sys/time.h>

int N = -1;               // input
unsigned T;               // time limit (seconds)
struct timeval stopTime;  // time when we should stop
unsigned long iterations; // amount of finished cycles of computing

unsigned long res; // current value

void start();
void reinit();
void calc(int);

int main (int argc, char ** argv) {
  if (argc != 2) {
    printf("Time limit parameter required:\n ./main.bin <time>\n");
  } else {
    iterations = 0;
    sscanf(argv[1], "%d", &T);
    printf("Calculating factorial of N many times during %u seconds\n", T);
    printf("Input N:\n");

    scanf("%d", &N);
    start();
  }
  return 0;
}

void start() {
  struct timeval now;
  gettimeofday(&stopTime, NULL);
  stopTime.tv_sec += T; // NOW + T seconds
  gettimeofday(&now, NULL);
  while (now.tv_sec <  stopTime.tv_sec ||
        (now.tv_sec == stopTime.tv_sec && now.tv_usec < stopTime.tv_usec)) {
    reinit();
    calc(N);
    iterations++;
    gettimeofday(&now, NULL);
  }
  iterations--; // last iteration started before time was up but finished after time was up
  // required '>' mark for automatic benchmark
  printf("> %lu iterations\n", iterations);
  printf("Resulting %d! = %lu\n",N,res);
}

void reinit() {
  res = 1L;
}

void calc(int N) {
  for (int i = N; i > 1; i--) {
    res *= i;
  }
}
